Firestore/Example/Tests/Util/FSTDispatchQueueTests.mm

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright 2018 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#import "Firestore/Source/Util/FSTDispatchQueue.h"
+
+#import <XCTest/XCTest.h>
+
+#import "Firestore/Example/Tests/Util/XCTestCase+Await.h"
+
+// We reuse these TimerIDs for generic testing.
+static const FSTTimerID timerID1 = FSTTimerIDListenStreamConnection;
+static const FSTTimerID timerID2 = FSTTimerIDListenStreamIdle;
+static const FSTTimerID timerID3 = FSTTimerIDWriteStreamConnection;
+
+@interface FSTDispatchQueueTests : XCTestCase
+@end
+
+@implementation FSTDispatchQueueTests {
+  FSTDispatchQueue *_queue;
+  NSMutableArray *_completedSteps;
+  NSArray *_expectedSteps;
+  XCTestExpectation *_expectation;
+}
+
+- (void)setUp {
+  [super setUp];
+  dispatch_queue_t dispatch_queue =
+      dispatch_queue_create("FSTDispatchQueueTests", DISPATCH_QUEUE_SERIAL);
+  _queue = [[FSTDispatchQueue alloc] initWithQueue:dispatch_queue];
+  _completedSteps = [NSMutableArray array];
+  _expectedSteps = nil;
+}
+
+/**
+ * Helper to return a block that adds @(n) to _completedSteps when run and fulfils _expectation if
+ * the _completedSteps match the _expectedSteps.
+ */
+- (void (^)())blockForStep:(int)n {
+  return ^void() {
+    [self->_completedSteps addObject:@(n)];
+    if (self->_expectedSteps && [self->_completedSteps isEqualToArray:self->_expectedSteps]) {
```

Comment:
This has the effect that any invalid outcome will hang the test.

We'll get better error messages if you release the expectation after the number of completed steps is >= expected steps and the assert the arrays are equal instead.

Reply:
Good point, done.